File: packages/flutter_code_push/android/src/main/java/dev/flutter/codepush/CodePushStorage.java
package dev.flutter.codepush;

import android.content.Context;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.util.Locale;
import org.json.JSONObject;

/** Stages and promotes OTA Dart isolate snapshot blobs (not native .so libraries). */
final class CodePushStorage {
  static final String ROOT_DIRECTORY_NAME = "code_push";
  static final String ACTIVE_DIRECTORY_NAME = "active";
  static final String STAGING_DIRECTORY_NAME = "staging";
  static final String MANIFEST_FILE_NAME = "patch_manifest.json";
  static final String ISOLATE_SNAPSHOT_DATA_FILE_NAME = "isolate_snapshot_data";
  static final String ISOLATE_SNAPSHOT_INSTR_FILE_NAME = "isolate_snapshot_instr";

  private final Context context;

  CodePushStorage(@NonNull Context context) {
    this.context = context.getApplicationContext();
  }

  @Nullable
  Integer readCurrentPatchNumber() throws IOException {
    File manifestFile = new File(getActiveDirectory(), MANIFEST_FILE_NAME);
    if (!manifestFile.exists()) {
      return null;
    }
    JSONObject manifest = readManifest(manifestFile);
    return manifest.has("patch_number") ? manifest.getInt("patch_number") : null;
  }

  void stagePatchFromUrls(
      int patchNumber,
      @NonNull String releaseVersion,
      @NonNull String dataDownloadUrl,
      @NonNull String instrDownloadUrl,
      @NonNull String dataSha256,
      @NonNull String instrSha256,
      @Nullable Long expectedDataLengthBytes,
      @Nullable Long expectedInstrLengthBytes,
      boolean enabled)
      throws IOException {
    File stagingDirectory = getStagingDirectory();
    deleteDirectoryContents(stagingDirectory);

    File dataFile = new File(stagingDirectory, ISOLATE_SNAPSHOT_DATA_FILE_NAME);
    File instrFile = new File(stagingDirectory, ISOLATE_SNAPSHOT_INSTR_FILE_NAME);
    downloadAndVerify(dataDownloadUrl, dataFile, dataSha256, expectedDataLengthBytes);
    downloadAndVerify(instrDownloadUrl, instrFile, instrSha256, expectedInstrLengthBytes);

    JSONObject manifest =
        new JSONObject()
            .put("patch_number", patchNumber)
            .put("release_version", releaseVersion)
            .put("isolate_data_sha256", dataSha256.toLowerCase(Locale.US))
            .put("isolate_instr_sha256", instrSha256.toLowerCase(Locale.US))
            .put("isolate_data_length_bytes", dataFile.length())
            .put("isolate_instr_length_bytes", instrFile.length())
            .put("enabled", enabled);
    writeManifest(new File(stagingDirectory, MANIFEST_FILE_NAME), manifest);
  }

  void applyStagedPatch() throws IOException {
    File stagingDirectory = getStagingDirectory();
    File stagedData = new File(stagingDirectory, ISOLATE_SNAPSHOT_DATA_FILE_NAME);
    File stagedInstr = new File(stagingDirectory, ISOLATE_SNAPSHOT_INSTR_FILE_NAME);
    File stagedManifest = new File(stagingDirectory, MANIFEST_FILE_NAME);
    if (!stagedData.exists() || !stagedInstr.exists() || !stagedManifest.exists()) {
      throw new IOException("No staged code push patch is available.");
    }

    File activeDirectory = getActiveDirectory();
    deleteDirectoryContents(activeDirectory);

    copyFile(stagedData, new File(activeDirectory, ISOLATE_SNAPSHOT_DATA_FILE_NAME));
    copyFile(stagedInstr, new File(activeDirectory, ISOLATE_SNAPSHOT_INSTR_FILE_NAME));
    copyFile(stagedManifest, new File(activeDirectory, MANIFEST_FILE_NAME));
    deleteDirectoryContents(stagingDirectory);
  }

  void clearActivePatch() {
    deleteDirectoryContents(getActiveDirectory());
    deleteDirectoryContents(getStagingDirectory());
  }

  private void downloadAndVerify(
      @NonNull String downloadUrl,
      @NonNull File destination,
      @NonNull String expectedSha256,
      @Nullable Long expectedLengthBytes)
      throws IOException {
    HttpURLConnection connection = null;
    try {
      connection = (HttpURLConnection) new URL(downloadUrl).openConnection();
      connection.setConnectTimeout(30_000);
      connection.setReadTimeout(300_000);
      connection.connect();
      if (connection.getResponseCode() != HttpURLConnection.HTTP_OK) {
        throw new IOException("Patch download failed with HTTP " + connection.getResponseCode());
      }

      MessageDigest digest = MessageDigest.getInstance("SHA-256");
      try (InputStream inputStream = connection.getInputStream();
          FileOutputStream outputStream = new FileOutputStream(destination)) {
        byte[] buffer = new byte[8192];
        int read;
        long totalBytes = 0;
        while ((read = inputStream.read(buffer)) != -1) {
          digest.update(buffer, 0, read);
          outputStream.write(buffer, 0, read);
          totalBytes += read;
        }
        if (expectedLengthBytes != null && totalBytes != expectedLengthBytes) {
          throw new IOException(
              "Patch size mismatch. expected="
                  + expectedLengthBytes
                  + " actual="
                  + totalBytes);
        }
      }

      String actualSha256 = toHex(digest.digest());
      if (!actualSha256.equalsIgnoreCase(expectedSha256)) {
        destination.delete();
        throw new IOException("Patch SHA-256 mismatch.");
      }
    } catch (Exception exception) {
      destination.delete();
      if (exception instanceof IOException) {
        throw (IOException) exception;
      }
      throw new IOException(exception);
    } finally {
      if (connection != null) {
        connection.disconnect();
      }
    }
  }

  private File getRootDirectory() {
    return new File(context.getFilesDir(), ROOT_DIRECTORY_NAME);
  }

  private File getActiveDirectory() {
    File directory = new File(getRootDirectory(), ACTIVE_DIRECTORY_NAME);
    directory.mkdirs();
    return directory;
  }

  private File getStagingDirectory() {
    File directory = new File(getRootDirectory(), STAGING_DIRECTORY_NAME);
    directory.mkdirs();
    return directory;
  }

  private static JSONObject readManifest(@NonNull File manifestFile) throws IOException {
    try (FileInputStream inputStream = new FileInputStream(manifestFile)) {
      byte[] bytes = readAllBytes(inputStream);
      return new JSONObject(new String(bytes, StandardCharsets.UTF_8));
    } catch (org.json.JSONException exception) {
      throw new IOException(exception);
    }
  }

  private static void writeManifest(@NonNull File manifestFile, @NonNull JSONObject manifest)
      throws IOException {
    try (FileOutputStream outputStream = new FileOutputStream(manifestFile)) {
      outputStream.write(manifest.toString().getBytes(StandardCharsets.UTF_8));
    }
  }

  private static void copyFile(@NonNull File source, @NonNull File destination) throws IOException {
    try (FileInputStream inputStream = new FileInputStream(source);
        FileOutputStream outputStream = new FileOutputStream(destination)) {
      inputStream.transferTo(outputStream);
    }
  }

  private static void deleteDirectoryContents(@NonNull File directory) {
    if (!directory.exists()) {
      return;
    }
    File[] children = directory.listFiles();
    if (children == null) {
      return;
    }
    for (File child : children) {
      if (!child.delete()) {
        child.deleteOnExit();
      }
    }
  }

  private static byte[] readAllBytes(@NonNull FileInputStream inputStream) throws IOException {
    byte[] buffer = new byte[8192];
    int read;
    java.io.ByteArrayOutputStream outputStream = new java.io.ByteArrayOutputStream();
    while ((read = inputStream.read(buffer)) != -1) {
      outputStream.write(buffer, 0, read);
    }
    return outputStream.toByteArray();
  }

  private static String toHex(byte[] bytes) {
    StringBuilder builder = new StringBuilder(bytes.length * 2);
    for (byte value : bytes) {
      builder.append(String.format(Locale.US, "%02x", value));
    }
    return builder.toString();
  }
}

File: packages/flutter_code_push/ios/Classes/CodePushStorage.swift
import CommonCrypto
import Foundation

final class CodePushStorage {
  static let rootDirectoryName = "code_push"
  static let activeDirectoryName = "active"
  static let stagingDirectoryName = "staging"
  static let manifestFileName = "patch_manifest.json"
  static let isolateSnapshotDataFileName = "isolate_snapshot_data"
  static let isolateSnapshotInstrFileName = "isolate_snapshot_instr"

  private let fileManager = FileManager.default

  private var documentsDirectory: URL {
    fileManager.urls(for: .documentDirectory, in: .userDomainMask).first!
  }

  private var rootDirectory: URL {
    documentsDirectory.appendingPathComponent(Self.rootDirectoryName, isDirectory: true)
  }

  private var activeDirectory: URL {
    rootDirectory.appendingPathComponent(Self.activeDirectoryName, isDirectory: true)
  }

  private var stagingDirectory: URL {
    rootDirectory.appendingPathComponent(Self.stagingDirectoryName, isDirectory: true)
  }

  func readCurrentPatchNumber() throws -> Int? {
    let manifestURL = activeDirectory.appendingPathComponent(Self.manifestFileName)
    guard fileManager.fileExists(atPath: manifestURL.path) else {
      return nil
    }
    let data = try Data(contentsOf: manifestURL)
    let json = try JSONSerialization.jsonObject(with: data) as? [String: Any]
    return json?["patch_number"] as? Int
  }

  func stagePatchFromUrls(
    patchNumber: Int,
    releaseVersion: String,
    dataDownloadURL: String,
    instrDownloadURL: String,
    dataSha256: String,
    instrSha256: String,
    dataLengthBytes: Int?,
    instrLengthBytes: Int?,
    enabled: Bool
  ) throws {
    try ensureDirectory(stagingDirectory)
    try removeDirectoryContents(stagingDirectory)

    let dataURL = stagingDirectory.appendingPathComponent(Self.isolateSnapshotDataFileName)
    let instrURL = stagingDirectory.appendingPathComponent(Self.isolateSnapshotInstrFileName)
    try downloadAndVerify(
      urlString: dataDownloadURL,
      destination: dataURL,
      expectedSha256: dataSha256,
      expectedLengthBytes: dataLengthBytes
    )
    try downloadAndVerify(
      urlString: instrDownloadURL,
      destination: instrURL,
      expectedSha256: instrSha256,
      expectedLengthBytes: instrLengthBytes
    )

    let manifest: [String: Any] = [
      "patch_number": patchNumber,
      "release_version": releaseVersion,
      "isolate_data_sha256": dataSha256.lowercased(),
      "isolate_instr_sha256": instrSha256.lowercased(),
      "isolate_data_length_bytes": try fileSize(dataURL),
      "isolate_instr_length_bytes": try fileSize(instrURL),
      "enabled": enabled,
    ]
    let manifestData = try JSONSerialization.data(withJSONObject: manifest)
    try manifestData.write(to: stagingDirectory.appendingPathComponent(Self.manifestFileName))
  }

  func applyStagedPatch() throws {
    let stagedData = stagingDirectory.appendingPathComponent(Self.isolateSnapshotDataFileName)
    let stagedInstr = stagingDirectory.appendingPathComponent(Self.isolateSnapshotInstrFileName)
    let stagedManifest = stagingDirectory.appendingPathComponent(Self.manifestFileName)
    guard fileManager.fileExists(atPath: stagedData.path),
      fileManager.fileExists(atPath: stagedInstr.path),
      fileManager.fileExists(atPath: stagedManifest.path)
    else {
      throw NSError(
        domain: "CodePushStorage",
        code: 1,
        userInfo: [NSLocalizedDescriptionKey: "No staged code push patch is available."]
      )
    }

    try ensureDirectory(activeDirectory)
    try removeDirectoryContents(activeDirectory)

    try fileManager.copyItem(at: stagedData, to: activeDirectory.appendingPathComponent(Self.isolateSnapshotDataFileName))
    try fileManager.copyItem(at: stagedInstr, to: activeDirectory.appendingPathComponent(Self.isolateSnapshotInstrFileName))
    try fileManager.copyItem(at: stagedManifest, to: activeDirectory.appendingPathComponent(Self.manifestFileName))
    try removeDirectoryContents(stagingDirectory)
  }

  func clearActivePatch() throws {
    try removeDirectoryContents(activeDirectory)
    try removeDirectoryContents(stagingDirectory)
  }

  private func downloadAndVerify(
    urlString: String,
    destination: URL,
    expectedSha256: String,
    expectedLengthBytes: Int?
  ) throws {
    guard let url = URL(string: urlString) else {
      throw NSError(
        domain: "CodePushStorage",
        code: 2,
        userInfo: [NSLocalizedDescriptionKey: "Invalid download URL."]
      )
    }

    var request = URLRequest(url: url)
    request.timeoutInterval = 300
    let (tempURL, response) = try URLSession.shared.download(for: request)
    defer { try? fileManager.removeItem(at: tempURL) }

    if let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode != 200 {
      throw NSError(
        domain: "CodePushStorage",
        code: 3,
        userInfo: [NSLocalizedDescriptionKey: "Patch download failed with HTTP \(httpResponse.statusCode)"]
      )
    }

    if fileManager.fileExists(atPath: destination.path) {
      try fileManager.removeItem(at: destination)
    }
    try fileManager.moveItem(at: tempURL, to: destination)

    let actualLength = try fileSize(destination)
    if let expectedLengthBytes, actualLength != expectedLengthBytes {
      try fileManager.removeItem(at: destination)
      throw NSError(
        domain: "CodePushStorage",
        code: 4,
        userInfo: [NSLocalizedDescriptionKey: "Patch size mismatch."]
      )
    }

    let actualSha256 = try sha256Hex(of: destination)
    if actualSha256.caseInsensitiveCompare(expectedSha256) != .orderedSame {
      try fileManager.removeItem(at: destination)
      throw NSError(
        domain: "CodePushStorage",
        code: 5,
        userInfo: [NSLocalizedDescriptionKey: "Patch SHA-256 mismatch."]
      )
    }
  }

  private func sha256Hex(of fileURL: URL) throws -> String {
    let handle = try FileHandle(forReadingFrom: fileURL)
    defer { try? handle.close() }

    var context = CC_SHA256_CTX()
    CC_SHA256_Init(&context)

    while true {
      let chunk = try handle.read(upToCount: 8192) ?? Data()
      if chunk.isEmpty {
        break
      }
      chunk.withUnsafeBytes { buffer in
        _ = CC_SHA256_Update(&context, buffer.baseAddress, CC_LONG(chunk.count))
      }
    }

    var digest = [UInt8](repeating: 0, count: Int(CC_SHA256_DIGEST_LENGTH))
    CC_SHA256_Final(&digest, &context)
    return digest.map { String(format: "%02x", $0) }.joined()
  }

  private func fileSize(_ url: URL) throws -> Int {
    let attributes = try fileManager.attributesOfItem(atPath: url.path)
    return attributes[.size] as? Int ?? 0
  }

  private func ensureDirectory(_ url: URL) throws {
    try fileManager.createDirectory(at: url, withIntermediateDirectories: true)
  }

  private func removeDirectoryContents(_ directory: URL) throws {
    guard fileManager.fileExists(atPath: directory.path) else {
      return
    }
    let children = try fileManager.contentsOfDirectory(at: directory, includingPropertiesForKeys: nil)
    for child in children {
      try fileManager.removeItem(at: child)
    }
  }
}
